Adding postgresql policies

SebastianClaesson opened the pull request:
# PostgreSQL policies

Adding new Azure policies for PostgreSQL.

- Azure policy to limit what versions are allowed
- Azure policy to deny the use of local postgreSQL admin user
- Azure policy to deny the use of password auth
- Azure policy to enforce a naming pattern when using Entra groups as Administrators
- Azure policy to enforce certain principal types when using Entra

SebastianClaesson commented:
Example result for "PostgreSQL flexible servers administrator - Enforce only specific administrator principal types" with parameter set to "Group" ;
![image](https://github.com/user-attachments/assets/9ce39519-d5c1-480e-87cb-7da7e6733133)
Non-compliant resource;
![image](https://github.com/user-attachments/assets/fd897a9c-483b-4bad-82d0-48d16c82dc1c)
Compliant resource;
![image](https://github.com/user-attachments/assets/16364d45-ddec-4511-8f87-7e59ad328e79)

aschabus commented:
@SebastianClaesson could you please change the category to "SQL"? This is the same category that the built-in PostgreSQL policies use.

SebastianClaesson commented:
> @SebastianClaesson could you please change the category to "SQL"? This is the same category that the built-in PostgreSQL policies use.

Sure thing, My opinion is that it should be split between SQL engines (mysql, postgresql/mssql etc.)
When more policies will be contributed, it would make sense that there was a folder for each resource provider to avoid confusion!

However, it's been updated with category and the folder "postgreSQL" has been removed!

SebastianClaesson commented:
Saw I still had "Custom" as policy type. It's been removed! :)

aschabus commented:
@SebastianClaesson thanks a lot

SebastianClaesson commented:
> @SebastianClaesson thanks a lot

Awesome, let me know if there's anything else! 🚀

aschabus commented:
Hi @SebastianClaesson - no, we just hadn't had a review meeting for some time.
But we should be able to look at it soom.

donk-msft commented:
I tested the policies and they work as expected, though non-compliance is not detected before deployment (but that's not related to the policy definitions). The 'Azure policy to deny the use of password auth' will prevent deployment of a PostgreSQL Flexible server through Azure Portal.

donk-msft commented:
Hi @SebastianClaesson,

You can remove the two definitions from this PR and we can approve it for the other three definitions. If and when you've solved the issue, you can raise another PR. If you don't agree, you can cancel the PR as you suggested.

SebastianClaesson commented:
I've updated the policy to match https://www.azadvertizer.net/azpolicyadvertizer/fa498b91-8a7e-4710-9578-da944c68d1fe.html#:~:text=Disabling%20local%20authentication%20methods%20and%20allowing%20only%20Microsoft,each%20column%20to%20filter%20and%20limit%20table%20data.

However, this policy did not allow the "deny" action, so I've basically added it and removed the other two policies.

I also have opened a ticket to Microsoft support as the Azure portal template is incorrect.
The issue:
```
"administratorLoginPassword": "[if(empty(parameters('administratorLoginPassword')), json('null'), parameters('administratorLoginPassword'))]",
```

Which returns {}, instead of an empty string, which breaks the Azure policy I had written that check for an empty string (seems the Azure policy engine gets a violation)
When explicitly set to null then it worked, however this policy also covers all api versions as it checks for "properties.authConfig.activeDirectoryAuth" which does not exist in the older apis.

![image](https://github.com/user-attachments/assets/9fc612d2-009b-416b-a9b2-8a0c5c69edbb)

Reference; https://learn.microsoft.com/en-us/azure/templates/microsoft.dbforpostgresql/flexibleservers?pivots=deployment-language-bicep#serverproperties

Now all the policies should be working as expected, I've tested them in my environment. 
However the validation for "enforce-postgresql-entra-only-authentication" will only occur after posting the template in the Azure portal and not on the basic blade.

donk-msft reviewed (approved):
@SebastianClaesson, thank you for looking into this. I approved the PR.
